pkg/reconciler/broker/broker.go

```diff
@@ -0,0 +1,480 @@
+		return fmt.Errorf("failed to reconcile triggers: %v", err)
+	}
+
+	logging.FromContext(ctx).Desugar().Info("targetsConfig", zap.Any("cfg", r.targetsConfig.String()))
+
+	return newReconciledNormal(b.Namespace, b.Name)
+}
+
+func (r *Reconciler) FinalizeKind(ctx context.Context, b *brokerv1beta1.Broker) pkgreconciler.Event {
+	logger := logging.FromContext(ctx).Desugar()
+	logger.Debug("Finalizing Broker")
+
+	// Reconcile triggers so they update their status
+	if err := r.reconcileTriggers(ctx, b); err != nil {
+		logging.FromContext(ctx).Desugar().Error("Problem reconciling triggers", zap.Error(err))
+		return fmt.Errorf("failed to reconcile triggers: %v", err)
+	}
+
+	if err := r.deleteDecouplingTopicAndSubscription(ctx, b); err != nil {
+		return fmt.Errorf("Failed to delete Pub/Sub topic: %v", err)
+	}
+
+	// TODO should we also delete the broker from the config? Maybe better
+	// to keep the targets in place if there are still triggers. But we can
+	// update the status to UNKNOWN and remove address etc. Maybe need a new
+	// status DELETED or a deleted timestamp that can be used to clean up later.
+
+	return newReconciledNormal(b.Namespace, b.Name)
+}
+
+func (r *Reconciler) reconcileBroker(ctx context.Context, b *brokerv1beta1.Broker) error {
+	logger := logging.FromContext(ctx).Desugar()
+	logger.Debug("Reconciling Broker", zap.Any("broker", b))
+	b.Status.InitializeConditions()
+	b.Status.ObservedGeneration = b.Generation
+
+	// Create decoupling topic and pullsub for this broker. Ingress will push
+	// to this topic and fanout will pull from the pull sub.
+	if err := r.reconcileDecouplingTopicAndSubscription(ctx, b); err != nil {
+		return fmt.Errorf("Decoupling topic reconcile failed: %w", err)
+	}
+
+	r.targetsConfig.MutateBroker(b.Namespace, b.Name, func(m config.BrokerMutation) {
+		m.SetID(string(b.UID))
+		m.SetAddress("") //TODO
+		m.SetDecoupleQueue(&config.Queue{
+			//TODO should this use the status fields or the name generator funcs?
+			Topic:        b.Status.TopicID,
+			Subscription: b.Status.SubscriptionID,
+		})
+		if b.Status.IsReady() {
+			m.SetState(config.State_READY)
+		} else {
+			m.SetState(config.State_UNKNOWN)
+		}
+	})
+
+	// Update config map
+	r.flagTargetsForUpdate()
+
+	//TODO configmap cleanup: if any brokers are in deleted state with no triggers
+	// (or all triggers are in deleted state), remove that entry
+	return nil
+}
+
+func (r *Reconciler) reconcileDecouplingTopicAndSubscription(ctx context.Context, b *brokerv1beta1.Broker) error {
+	logger := logging.FromContext(ctx).Desugar()
+	logger.Debug("Reconciling decoupling topic")
+	// get ProjectID from config or metadata
+	//TODO(grantr) support configuring project in broker config
+	projectID, err := utils.ProjectID("")
+	if err != nil {
+		logger.Error("Failed to find project id", zap.Error(err))
+		return err
+	}
+	// Set the projectID in the status.
+	b.Status.ProjectID = projectID
```

Comment:
Probably not. This would get really complicated if brokers shared a data plane but could have different projects. If project is configurable it should probably be a property of the data plane rather than the broker.

I still think it's useful to have the project id here as a debugging signal to the user. What do you think?

Reply:
It has some value for debugging. It's up to you to keep it or not :)